src/librustc/traits/select.rs

```diff
@@ -3381,21 +3397,20 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 // with what our behavior should be there. -nikomatsakis
                 let InferOk { obligations, .. } = self.infcx
                     .at(&obligation.cause, obligation.param_env)
-                    .eq(target, source_trait) // FIXME -- see below
+                    .eq(target, source_ty) // FIXME -- see above
                     .map_err(|_| Unimplemented)?;
                 nested.extend(obligations);
+                */
 
-                // Register one obligation for 'a: 'b.
-                let cause = ObligationCause::new(
-                    obligation.cause.span,
-                    obligation.cause.body_id,
-                    ObjectCastObligation(target),
+                // Register an obligation for `dyn TraitA: TraitB`.
+                nested.extend(
+                    data_b.iter()
+                        .map(|d| predicate_to_obligation(d.with_self_ty(tcx, source_ty))),
                 );
-                let outlives = ty::OutlivesPredicate(r_a, r_b);
-                nested.push(Obligation::with_depth(
-                    cause,
-                    obligation.recursion_depth + 1,
-                    obligation.param_env,
+
+                // Register an obligation for `'a: 'b`.
+                let outlives = ty::OutlivesPredicate(region_a, region_b);
+                nested.push(predicate_to_obligation(
```

Comment:
Nit: seems like you lost the `recursion_depth + 1` count here.

Reply:
Have I actually? `predicate_to_obligation` does `recursion_depth + 1` so I this is okay... I think?